source/merge.ts

```diff
@@ -38,42 +44,34 @@ export default function merge<Target extends {[key: string]: unknown}, Source ex
 	return target as Target & Source;
 }
 
-export function mergeOptions(...sources: Array<Partial<Options>>): Partial<Options> & {hooks: Partial<Hooks>} {
-	sources = sources.map(source => source || {});
-	const merged = merge({}, ...sources);
+export function mergeOptions<T extends Options>(...sources: T[]): T & { hooks: Partial<Hooks> } {
+	const mergedOptions = merge({} as T & { hooks: Partial<Hooks> }, ...sources.map(source => source || {}));
 
-	// TODO: This is a funky situation. Even though we "know" that we're going to
-	//       populate the `hooks` object in the loop below, TypeScript want us to
-	//       put them into the object upon initialization, because it cannot infer
-	//       that they are going to conform correctly in runtime.
-	const hooks = {} as {[Key in HookEvent]: HookType[]};
-	for (const hook of knownHookEvents) {
-		hooks[hook] = [];
-	}
+	const hooks = knownHookEvents.reduce((acc, current) => ({...acc, [current]: []}), {}) as Record<HookEvent, HookType[]>;
 
 	for (const source of sources) {
-		if (source.hooks) {
+		if (source && source.hooks) {
 			for (const hook of knownHookEvents) {
 				hooks[hook] = hooks[hook].concat(source.hooks[hook] || []);
 			}
 		}
 	}
 
-	merged.hooks = hooks as Hooks;
+	mergedOptions.hooks = hooks as Hooks;
 
-	return merged as Partial<Options> & {hooks: Partial<Hooks>};
+	return mergedOptions;
 }
 
-export function mergeInstances(instances: InterfaceWithDefaults[], methods?: Method[]): Instance {
+export function mergeInstances(instances: Got[], methods?: Method[]): Defaults {
 	const handlers = instances.map(instance => instance.defaults.handler);
 	const size = instances.length - 1;
 
 	return {
 		methods,
-		options: mergeOptions(...instances.map(instance => instance.defaults.options)),
-		handler: (options: Options, next: NextFunction) => {
-			let iteration = -1;
-			const iterate = (options: Options): void => handlers[++iteration](options, iteration === size ? next : iterate);
+		options: mergeOptions(...instances.map(instance => instance.defaults.options || {})),
+		handler: <T extends ProxyStream | CancelableRequest<Response>>(options: NormalizedOptions, next: (options: NormalizedOptions) => T) => {
+			let iteration = 0;
+			const iterate = (newOptions: NormalizedOptions): T => handlers[iteration++]!(newOptions, iteration === size ? next : iterate);
```

Comment:
This is also incorrect. If you use `iteration++` you will miss the last handler.